### PR TITLE
Convert thread and future enums to class enums

### PR DESCRIPTION
**Pull request by msimberg**
Partially addresses #4702.

### CODE REVIEW

File: components/iostreams/src/server/output_stream.cpp [hunk 1]
```diff
@@ -81,7 +81,7 @@ namespace hpx { namespace iostreams { namespace server
         pending_output_.output(locality_id, count, in, write_f, mtx_);
 
         // Wake up caller.
-        threads::set_thread_state(caller, threads::pending);
+        threads::set_thread_state(caller, threads::thread_state_enum::pending);
```
Comment (hkaiser) [inline components/iostreams/src/server/output_stream.cpp:84]: Would it be appropriate to use this opportunity to rename the enum to `hpx::threads::thread_state`?

**Review by hkaiser — DISMISSED** [review]
LGTM, thanks for looking into this!

File: components/iostreams/src/server/output_stream.cpp (hunk 1 above)
Comment (msimberg) [reply components/iostreams/src/server/output_stream.cpp:84]: That's what I wanted to do first :P we already have a `thread_state` though, which is the combination of `thread_state_enum` and `thread_state_ex_enum` (I was very confused why they'd need `enum` in the name until I remembered that): https://github.com/STEllAR-GROUP/hpx/blob/e907e3227dd0b22e4e3c7c662a608490de8adb5b/libs/core/coroutines/include/hpx/coroutines/thread_enums.hpp#L139-L141.

I think the combined state is probably the rightful owner of the `thread_state` name (not very strongly, I'm open to alternatives here). That said, I realize now that `thread_state_ex_enum` definitely doesn't need the `enum` except for symmetry with the other enum. The main enum could be `thread_schedule_state`? And the extended one `thread_wait_state`? Or `thread_restart_state`? Or something else?

In any case, a rename would indeed be appropriate.

Comment (hkaiser) [reply components/iostreams/src/server/output_stream.cpp:84]: > That's what I wanted to do first :P we already have a `thread_state` though, which is the combination of `thread_state_enum` and `thread_state_ex_enum` (I was very confused why they'd need `enum` in the name until I remembered that):
> 
> https://github.com/STEllAR-GROUP/hpx/blob/e907e3227dd0b22e4e3c7c662a608490de8adb5b/libs/core/coroutines/include/hpx/coroutines/thread_enums.hpp#L139-L141
> 
> .
> I think the combined state is probably the rightful owner of the `thread_state` name (not very strongly, I'm open to alternatives here). That said, I realize now that `thread_state_ex_enum` definitely doesn't need the `enum` except for symmetry with the other enum. The main enum could be `thread_schedule_state`? And the extended one `thread_wait_state`? Or `thread_restart_state`? Or something else?
> 
> In any case, a rename would indeed be appropriate.

Fair points. I withdraw my suggestion. Yours are acceptable.

**Review by aurianer — DISMISSED** [review]
LGTM thanks a lot!

**Comment by msimberg** [comment]
@hkaiser, @K-ballo, any idea what MSVC is not happy about here: https://github.com/STEllAR-GROUP/hpx/pull/5046/checks?check_run_id=1374461029?

**Comment by K-ballo** [comment]
> @hkaiser, @K-ballo, any idea what MSVC is not happy about here: https://github.com/STEllAR-GROUP/hpx/pull/5046/checks?check_run_id=1374461029?

Windows SDK defines `small` as `char`.

File: components/iostreams/src/server/output_stream.cpp (hunk 1 above)
Comment (msimberg) [reply components/iostreams/src/server/output_stream.cpp]: I updated these to be `thread_schedule_state` and `thread_wait_state` now.

**Comment by msimberg** [comment]
So, I've renamed `thread_stacksize::small` to `thread_stacksize::small_`. I don't particularly like it but we have precedent with `default_`. I like it more than alternative names (`little`, `tiny`?) though, which is why I've kept it like this. Can someone think of another better name, or another workaround?

**Review by hkaiser — APPROVED** [review]
LGTM, thanks!

One minor comment (doesn't need to go in here): we should add `small` to the list of identifiers not to use (similar to the apple macros we are flagging as prohibited: https://github.com/STEllAR-GROUP/hpx/blob/master/tools/inspect/apple_macro_check.cpp#L27)

**Comment by msimberg** [comment]
> LGTM, thanks!
> 
> One minor comment (doesn't need to go in here): we should add `small` to the list of identifiers not to use (similar to the apple macros we are flagging as prohibited: https://github.com/STEllAR-GROUP/hpx/blob/master/tools/inspect/apple_macro_check.cpp#L27)

Good suggestion! I'll add it here, otherwise it'll never be done...

**Review by hkaiser — APPROVED** [review]
LGTM, thanks!

**Pull request merged**